crates/sui-core/src/authority_active/checkpoint_executor/mod.rs

```diff
@@ -0,0 +1,206 @@
+        let highest_executed_checkpoint =
+            Box::new(checkpoint_store.get_highest_executed_checkpoint_seq_number()?);
+
+        Ok(Self {
+            mailbox,
+            checkpoint_store,
+            highest_executed_checkpoint,
+            authority_state,
+            metrics,
+        })
+    }
+
+    pub async fn run(mut self) {
+        loop {
+            match self.mailbox.recv().await {
+                Ok(checkpoint) => {
+                    let last_to_exec = checkpoint.sequence_number() as u64;
+                    let next_to_exec = self
+                        .highest_executed_checkpoint
+                        .map(|highest| highest + 1)
+                        .unwrap_or(0);
+
+                    for seq_num in next_to_exec..=last_to_exec {
+                        let next_checkpoint = match self
+                            .checkpoint_store
+                            .get_checkpoint_by_sequence_number(seq_num)
+                        {
+                            Ok(Some(new_checkpoint)) => new_checkpoint,
+                            Ok(None) => {
+                                error!(
+                                    "Checkpoint {:?} does not exist in checkpoint store",
+                                    seq_num,
+                                );
+                                self.metrics.checkpoint_exec_errors.inc();
+                                continue;
+                            }
+                            Err(err) => {
+                                error!("Failed to fetch checkpoint {:?}: {:?}", seq_num, err);
+                                self.metrics.checkpoint_exec_errors.inc();
+                                continue;
+                            }
+                        };
+
+                        // Since StateSync is guaranteed to produce checkpoint sync messages
+                        // in monotonically increasing order, then a mismatch of epoch would
+                        // mean that we failed to execute reconfig at the epoch boundary.
+                        assert!(
+                            next_checkpoint.epoch()
+                                != self
+                                    .authority_state
+                                    .committee_store()
+                                    .get_latest_committee()
+                                    .epoch()
+                        );
+
+                        while let Err(err) = self.execute_checkpoint(&next_checkpoint).await {
+                            error!(
+                                "Error while executing checkpoint, will retry in 1s: {:?}",
+                                err
+                            );
+                            tokio::time::sleep(Duration::from_secs(1)).await;
+                            self.metrics.checkpoint_exec_errors.inc();
+                            continue;
+                        }
```

Comment:
Well transaction execution is done in a retry loop with a max attempts of 10 currently, so I suppose it could be argued that retying from this level isn't too useful..

Reply:
Going to leave this here for now until we've decided how best to handle it. I don't think this should happen hardly at all given transaction retries, but also don't see the harm in retrying (for example, in case reading from the store fails)

Reply:
I don't think we need to worry about this too much - execution errors should eventually become `assert!`s, as execution cannot fail on a healthy node, unless there are missing dependencies. And since all executions should go through transaction manager, it will make sure we never attempt execution until all deps are present.